mitmproxy/proxy/mode_specs.py

```diff
@@ -0,0 +1,199 @@
+"""
+This module is responsible for parsing proxy mode specifications such as
+`"regular"`, `"reverse:https://example.com"`, or `"socks5@1234"`. The general syntax is
+
+    mode [: mode_configuration] [@ [listen_addr:]listen_port]
+
+For a full example, consider `reverse:https://example.com@127.0.0.1:443`.
+This would spawn a reverse proxy on port 443 bound to localhost.
+The mode is `reverse`, and the mode data is `https://example.com`.
+Examples:
+
+    mode = ProxyMode.parse("regular@1234")
+    assert mode.listen_port == 1234
+    assert isinstance(mode, RegularMode)
+
+    ProxyMode.parse("reverse:example.com@invalid-port")  # ValueError
+
+"""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from functools import cache
+from typing import ClassVar, Literal
+
+from mitmproxy.coretypes.serializable import Serializable
+from mitmproxy.net import server_spec
+
+
+@dataclass(frozen=True)  # type: ignore
+class ProxyMode(Serializable, metaclass=ABCMeta):
+    """
+    Parsed representation of a proxy mode spec. Subclassed for each specific mode,
+    which then does its own data validation.
+    """
+    full_spec: str
+    data: str
+    custom_listen_host: str | None
+    custom_listen_port: int | None
+
+    transport_protocol: ClassVar[Literal["tcp", "udp"]] = "tcp"
+    """
+    The transport protocol used by this mode. Used to detect multiple servers targeting the same proto+port.
+    """
+    default_port: ClassVar[int] = 8080
+    __modes: ClassVar[dict[str, type[ProxyMode]]] = {}
+
+    @abstractmethod
+    def __post_init__(self) -> None:
+        """Validation of data happens here."""
+
+    def listen_host(self) -> str:
```

Comment:
Nit: Is there a reason why `listen_host` and `listen_port` aren't marked `@property` like `type`? Since `transport_protocol` can also be accesses without `()`, it might be more unified.

Reply:
They now take an argument to make it explicit that they depend on `ctx.master.options`.